sound/soc/sof/nocodec.c

```diff
@@ -59,8 +66,40 @@ static int sof_nocodec_bes_setup(struct device *dev,
 			links[i].dpcm_capture = 1;
 	}
 
+	/* add 2 dmic, as link_num will change on different platform hardcode
+	 *dmic be dai id to 99 and 100
+	 */
+	for (i = link_num; i < link_num + 2; i++) {
+		dlc = devm_kzalloc(dev, 2 * sizeof(*dlc), GFP_KERNEL);
+		if (!dlc)
+			return -ENOMEM;
+
+		links[i].cpus = &dlc[0];
+		links[i].codecs = dmic_component;
+		links[i].platforms = &dlc[1];
+
+		if (i == link_num) {
+			/* add dmic01 with id 99 */
+			links[i].id = 99;
```

Comment:
Should we use a define for dmic id? eg. ``#define DMIC_ID_OFFSET 99``

Reply:
no, you should use the next available link id, or something reasonable.
You are confusing PCM number and BE number.

Reply:
also no it's not good, we don't want DMIC for legacy platforms without dmic.

Reply:
@plbossart shoud we check the NHLT for DMIC info and add them here if they're present?

Reply:
it depends on what the objective is. NHTL w/ DMIC is only used for HDaudio platforms, this is not used on any Chromebook and recent UpExtreme.
I would think we add a BE dailink for all interfaces supported by the SOC.

Reply:
@plbossart My propose here is try to unify the BE dai codec name with other machine driver. So we can reuse the DMIC m4 macro in nocodec mode. No we are using **Nocodec-6** as the DMIC dai name

Why I use a static BE DAI ID is that the link_num is a changeable number
https://github.com/thesofproject/linux/blob/b42cc42e10c6300aa00e856850ad7aa279913909/sound/soc/sof/intel/hda.h#L358-L366
This makes impossible to write a valid toplogy.

I think only expose some BE DAI link like other machine drivers will do no harm for the platform. It is the TPLG to decided if we want to use it.

Reply:
How about adding DMIC DAIs to hda-dai.c, with that, we don't have to use the hard-coded 99 and 100 for DMIC?

Reply:
@keyonjie the DMIC DAIs are already included in hda-dai.c, see explicit references to DMIC01 Pin and DMIC16k Pin
Here the problem is how to declare dai-links in the machine driver.

Reply:
> I think only expose some BE DAI link like other machine drivers will do no harm for the platform. It is the TPLG to decided if we want to use it.

You cannot add a BE for DMIC if the DMIC DAI does not exist, e.g. on legacy platforms or i.MX.